### ISSUE TITLE
Softmax Decomp Causes Incorrect Gradients when Using `torch.compile` with `F.multi_head_attention_forward`

### ISSUE THREAD

**Issue by defaultd661** [pytorch/pytorch#152309]
### 🐛 Describe the bug

When using `torch.compile` to compile a model that internally calls `torch.nn.functional.multi_head_attention_forward`, the computed gradients differ significantly from the ones obtained via eager mode.

### To Reproduce
```
import torch
import torch.nn as nn
import torch.nn.functional as F


class ReproMultihead(nn.Module):

    def __init__(self):
        super().__init__()
        self.embed_dim = 256
        self.num_heads = self.embed_dim // 64
        self.in_proj_weight = nn.Parameter(torch.empty(3 * self.embed_dim,
            self.embed_dim))
        self.in_proj_bias = nn.Parameter(torch.empty(3 * self.embed_dim))
        self.out_proj_weight = nn.Parameter(torch.empty(self.embed_dim,
            self.embed_dim))
        self.out_proj_bias = nn.Parameter(torch.empty(self.embed_dim))
        nn.init.constant_(self.in_proj_weight, 0.1)
        nn.init.constant_(self.in_proj_bias, 0.1)
        nn.init.constant_(self.out_proj_weight, 0.1)
        nn.init.constant_(self.out_proj_bias, 0.1)

    def forward(self, x):
        x_t = x.transpose(0, 1)
        attn_output, _ = F.multi_head_attention_forward(query=x_t, key=x_t,
            value=x_t, embed_dim_to_check=self.embed_dim, num_heads=self.
            num_heads, in_proj_weight=self.in_proj_weight, in_proj_bias=
            self.in_proj_bias, bias_k=None, bias_v=None, add_zero_attn=
            False, dropout_p=0.0, out_proj_weight=self.out_proj_weight,
            out_proj_bias=self.out_proj_bias, training=True,
            key_padding_mask=None, need_weights=False, attn_mask=None,
            use_separate_proj_weight=False, q_proj_weight=None,
            k_proj_weight=None, v_proj_weight=None, static_k=None, static_v
            =None, average_attn_weights=True, is_causal=False)
        return attn_output.transpose(0, 1)

def test_bug():
    torch.set_default_device('cuda')
    torch.manual_seed(0)

    model = ReproMultihead().cuda()

    compiled_model = ReproMultihead().cuda()
    compiled_model = torch.compile(compiled_model)

    x = torch.randn((1, 512, 256), device='cuda', requires_grad=True)
    x_compiled = x.clone().detach().requires_grad_(True)

    out_eager = model(x)
    out_compiled = compiled_model(x_compiled)

    out_eager.sum().backward()
    out_compiled.sum().backward()

    weight_diff = torch.max(torch.abs(model.in_proj_weight.grad -
        compiled_model.in_proj_weight.grad)).item()
    print('weight_diff =', weight_diff)
    bias_diff = torch.max(torch.abs(model.in_proj_bias.grad -
        compiled_model.in_proj_bias.grad)).item()
    print('bias_diff =', bias_diff)

if __name__ == '__main__':
    test_bug()
```

### Output
```
weight_diff = 0.130126953125
bias_diff = 0.12890625
```

### Versions

PyTorch version: 2.7.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @chauhang @penguinwu @SherlockNoMad @bdhirsh

**Comment by eellison** [comment]
I couldn't repro this on a100 or h100.

```
weight_diff = 0.0
bias_diff = 0.0009765625
```

Can you try again on main, and reopen if this issue still exists ?

**Comment by eellison** [comment]
cc @drisspg can you repro ?

**Comment by eellison** [comment]
Okay, I can repro this by forcing the math backend. And the error is due to the softmax decomp. 


```
import torch
import torch.nn as nn
import torch.nn.functional as F


class ReproMultihead(nn.Module):

    def __init__(self):
        super().__init__()
        self.embed_dim = 256
        self.num_heads = self.embed_dim // 64
        self.in_proj_weight = nn.Parameter(torch.empty(3 * self.embed_dim,
            self.embed_dim))
        self.in_proj_bias = nn.Parameter(torch.empty(3 * self.embed_dim))
        self.out_proj_weight = nn.Parameter(torch.empty(self.embed_dim,
            self.embed_dim))
        self.out_proj_bias = nn.Parameter(torch.empty(self.embed_dim))
        nn.init.constant_(self.in_proj_weight, 0.1)
        nn.init.constant_(self.in_proj_bias, 0.1)
        nn.init.constant_(self.out_proj_weight, 0.1)
        nn.init.constant_(self.out_proj_bias, 0.1)

    def forward(self, x):
        x_t = x.transpose(0, 1)
        attn_output, _ = F.multi_head_attention_forward(query=x_t, key=x_t,
            value=x_t, embed_dim_to_check=self.embed_dim, num_heads=self.
            num_heads, in_proj_weight=self.in_proj_weight, in_proj_bias=
            self.in_proj_bias, bias_k=None, bias_v=None, add_zero_attn=
            False, dropout_p=0.0, out_proj_weight=self.out_proj_weight,
            out_proj_bias=self.out_proj_bias, training=True,
            key_padding_mask=None, need_weights=False, attn_mask=None,
            use_separate_proj_weight=False, q_proj_weight=None,
            k_proj_weight=None, v_proj_weight=None, static_k=None, static_v
            =None, average_attn_weights=True, is_causal=False)
        return attn_output.transpose(0, 1)

def test_bug():

    torch.backends.cuda.enable_flash_sdp(False)
    torch.backends.cuda.enable_mem_efficient_sdp(False)
    # torch.backends.cuda.enable_math_sdp(False)


    def test_fn():
        torch.set_default_device('cuda')
        torch.manual_seed(0)

        model = ReproMultihead().cuda()
        torch._dynamo.reset()
        compiled_model = ReproMultihead().cuda()
        compiled_model = torch.compile(compiled_model)

        x = torch.randn((1, 512, 256), device='cuda', requires_grad=True)
        x_compiled = x.clone().detach().requires_grad_(True)

        out_eager = model(x)
        out_compiled = compiled_model(x_compiled)

        out_eager.sum().backward()
        out_compiled.sum().backward()

        try:
            torch.testing.assert_close(model.in_proj_weight.grad, compiled_model.in_proj_weight.grad)
            return True
        except:
            return False
        
        # weight_diff = torch.max(torch.abs(model.in_proj_weight.grad -
        #     compiled_model.in_proj_weight.grad)).item()
        # print('weight_diff =', weight_diff)
        # bias_diff = torch.max(torch.abs(model.in_proj_bias.grad -
        #     compiled_model.in_proj_bias.grad)).item()
        # print('bias_diff =', bias_diff)

    from torch._inductor.compiler_bisector import CompilerBisector

    CompilerBisector.do_bisect(test_fn)



if __name__ == '__main__':
    test_bug()

```

```
No bisection status found.
Starting bisection process with system: eager
Moving to the next system: aot_eager
Moving to the next system: aot_eager_decomp_partition
The issue is in the aot_eager_decomp_partition system. Moving to the first subsystem: ConfigChange(name='aot_eager_decomp_partition_cse', config_name='aot_eager_decomp_partition', config_field='cse', config_value=False)
Disabling aot_eager_decomp_partition_cse did not fix the issue.
Moving to the next subsystem: aot_eager_decomp_partition - decomposition
Disabling decomposition fixed the issue.
Starting bisect by getting upper bound.
Upper bound of 76 found for aot_eager_decomp_partition.
Bisecting aot_eager_decomp_partition - decomposition (Range: [0, 76], Midpoint: 38)
Bisecting aot_eager_decomp_partition - decomposition (Range: [0, 38], Midpoint: 19)
Bisecting aot_eager_decomp_partition - decomposition (Range: [0, 19], Midpoint: 9)
Bisecting aot_eager_decomp_partition - decomposition (Range: [10, 19], Midpoint: 14)
Bisecting aot_eager_decomp_partition - decomposition (Range: [10, 14], Midpoint: 12)
Bisecting aot_eager_decomp_partition - decomposition (Range: [10, 12], Midpoint: 11)
Binary search completed for aot_eager_decomp_partition - decomposition. The bisect number is 12. Debug info: <OpOverload(op='aten._softmax', overload='default')>
Bisection status deleted.
```


@drisspg, any thoughts ?

**Comment by bdhirsh** [comment]
I have a pretty small repro of the softmax decomp causing noticeable differences compared to the eager kernel for `aten._softmax_backward_data`:
```
import torch
from torch._inductor import inductor_prims

# decomp from torch/_inductor/decompositions.py
def _softmax_backward_data_decomp(grad_output: torch.Tensor, output: torch.Tensor, dim: int, input_dtype: torch.dtype) -> torch.Tensor:
    new_grad_output = grad_output * output
    sum_new_grad = torch.sum(new_grad_output, dim=dim, keepdim=True)
    # non-FMA version is commented out
    # grad_input = new_grad_output - output * sum_new_grad
    grad_input = inductor_prims.fma(-output, sum_new_grad, new_grad_output)

    if grad_output.dtype != input_dtype:
        grad_input = grad_input.to(input_dtype)
    out = grad_input.contiguous()
    return out

grad_output = torch.randn(1, 4, 512, 512, device='cuda') * 1000
output = torch.rand(1, 4, 512, 512, device='cuda')

out_ref = torch.ops.aten._softmax_backward_data.default(grad_output, output, dim=-1, input_dtype=torch.float32)
out_test = _softmax_backward_data_decomp(grad_output, output, dim=-1, input_dtype=torch.float32)

print(torch.max(torch.abs(out_ref - out_test)))
```

A few things to call out:

(1) the difference in my tiny repro is `tensor(0.0078, device='cuda:0')`. which... is not huge, but seems like it is large enough to be problematic? in the full repro above, this small difference causes a larger difference in the final weight/bias grad values:
```
fw diff = 0.0008087158203125
weight_diff = 0.02783203125
bias_diff = 0.00390625
```


(2) all of the inputs/outputs are float32, so there don't seem to be any funky intermediate dtype casting differences going on.

(3) The decomp uses FMA. i tried both the FMA and non-FMA flavor, and I get similar divergences

(4) how large the difference is depends on the input data distribution. I got a similar data distribution to the MHA example in my repro above (`output` tensor is between 0 and 1`, `grad_output` is larger)

 I spent some time staring at the cuda kernel but so far I haven't figured out what causes the `_softmax_backward_data` cuda kernel ([link](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cuda/SoftMax.cu#L1410)) to diverge from the decomp.

cc @ngimel, in case you know

**Comment by ngimel** [comment]
It might be fma, cuda kernel might or might not produce fma instructions? Also torch.sum() and result produced in the kernel might slightly differ, as you said the error for the tiny repro is not huge so it's those small things. Softmax, as we know, is particularly sensitive to fma/no fma.

**Comment by bdhirsh** [comment]
It didn't appear to be FMA because i tried with FMA on and off and saw similar differences.

Are we ok chalking these differences up to reduction numerics being "different" between eager and compile and living with the (small) difference?

**Comment by bdhirsh** [comment]
One thing I will say is that there are two independent issues here:

(1) the softmax decomp causing slightly different numerics compared to eager (this can be fully repro'd with `backend="aot_eager_decomp_partition"`), that gives modest differences vs. eager in the final weight/bias grads:
```
weight_diff = 0.02783203125
bias_diff = 0.00390625
```

(2) there is a second issue I found that only repros with inductor, that I could repro with the efficient_attention backend, that causes larger differences:
```
weight_diff = 0.130126953125
bias_diff = 0.12890625
```

I can file a separate issue for that one

**Comment by ngimel** [comment]
@bdhirsh for your micro repro it's definitely just fp error, when I modify the output to
```
vals, indices = torch.max(torch.abs(out_ref.view(-1) - out_test.view(-1)), dim=0)
print(out_ref.view(-1)[indices], out_test.view(-1)[indices])
```
the output I'm getting is `tensor(-36137.9570, device='cuda:0') tensor(-36137.9453, device='cuda:0')`, the relative error is right where you'd expect it to be (1e-6)

**Comment by bdhirsh** [comment]
@ngimel doing the same calculation for the weight grad (instead of the forward out) gives me in the ballpark of `3e-5`: a bit larger, but not in a way that seems significant.

Do you have a standard set of utils that you recommend using for deciding if "compile vs eager delta is reasonable or not? I've been going off of `torch.max(torch.abs(test - ref))`, although I agree that this only gives absolute difference which doesn't tell the whole story.

I did this after reading your commend:
```
# prints tensor(-3.0756e-05, device='cuda:0')
vals, indices = torch.max(torch.abs(model.in_proj_weight.grad.view(-1) - compiled_model.in_proj_weight.grad.view(-1)), dim=0)
print(1 - (model.in_proj_weight.grad.view(-1)[indices] / compiled_model.in_proj_weight.grad.view(-1)[indices]))
```

**Comment by bdhirsh** [comment]
If Natalia / @eellison agree that the relative difference across compile vs eager here is within tolerance, i'll go ahead and close this issue

**Comment by ngimel** [comment]
@bdhirsh in the torch.compile torchbench test suite we are computing reference in fp64, and then check that difference (in the cosine similarity sense) from this gold standard is approximately the same for eager and compile, but at times it can hide real issues. Having a test with perfect accuracy and no false positives is hard

**Comment by bdhirsh** [comment]
tentatively closing this issue
